### DISCUSSION TITLE
Local time and timezone support for MicroPython (meta-discussion)

### DISCUSSION THREAD

**Discussion by projectgus** [micropython/micropython#12378] [Ideas]
Summarising this common problem and the current best solutions *and* possibility for a long-term fix.

## Problem

Some embedded applications care about the local time. MicroPython always uses UTC timezone and its [time](https://docs.micropython.org/en/latest/library/time.html) module is generally not timezone or localtime aware.

This seems to usually impact folks using [ntptime module from micropython-lib](https://github.com/micropython/micropython-lib/blob/master/micropython/net/ntptime/ntptime.py). This returns UTC (NTP protocol only uses UTC). How to display a local time somewhere?

## Current Simplest Solutions :heavy_check_mark: 

* `mip install datetime` from micropython-lib and [convert the timestamp as shown here](https://github.com/orgs/micropython/discussions/9545#discussioncomment-3822001) (thanks @jimmo).
* [Make your own one-liner `localtime()` function](https://github.com/orgs/micropython/discussions/11173#discussioncomment-6876888), as demonstrated by @MrEbbinghaus (thanks!).

## Other Proposed Solutions

* @bellingeri has a [pure Python POSIX timezone parser](https://github.com/bellingeri/localPTZtime) (impressive!). See also the [discussion about it](https://github.com/orgs/micropython/discussions/11150).

The following solutions add timezone offset support to the ntptime module:

* @peterhinch has an [alternative ntptime.py in his micropython-samples](https://github.com/peterhinch/micropython-samples/blob/master/ntptime/ntptime.py).
* @MrTinkerman has [another ntptime.py here](https://github.com/orgs/micropython/discussions/11173).
* Multiple PRs have proposed adding local time into micropython-lib ntptime:
   - https://github.com/micropython/micropython-lib/pull/715
   - https://github.com/micropython/micropython-lib/pull/597
   - https://github.com/micropython/micropython-lib/pull/635
   - https://github.com/micropython/micropython-lib/pull/587
   - https://github.com/micropython/micropython-lib/pull/695
   - https://github.com/micropython/micropython/pull/7075

This is clearly something people run into a lot!

## Long term solutions

As others, including maintainers, have already said in the linked discussions, the official micropython-lib `ntptime` module should keep working with UTC. However, it would clearly be useful to have an easy way to convert timestamps to a local timezone in MicroPython.

There are a range of possible solutions here, from a simple "local time offset" variable to a full [POSIX timezone format](https://www.gnu.org/software/libc/manual/html_node/TZ-Variable.html) compliant implementation that automatically switches for DST, etc.

## Current Status

* Except for the Unix port, the [time module](https://docs.micropython.org/en/latest/library/time.html) in MicroPython has no local time. `time.gmtime()` and `time.localtime()` are literally the same function.
* The result of `time.gmtime()` and `time.localtime()` [in MicroPython](https://docs.micropython.org/en/latest/library/time.html#time.localtime) is different [to CPython](https://docs.python.org/3.5/library/time.html#time.struct_time). MicroPython returns a tuple with 8 of the 9 indexed fields found in CPython `struct_time`. CPython's result has a ninth `tm_isdst` field, plus named access to those 9 fields and the extra fields `tm_zone` and `tm_gmtoff`.
* None of CPython's [time.timezone, time.tzname, time.tzset(), etc.](https://docs.python.org/3.5/library/time.html#time.timezone) are implemented in MicroPython due to (reasonable) size constraints.
* [micropython-lib datetime](https://github.com/micropython/micropython-lib/tree/master/python-stdlib/datetime) is "micro" in size but very comprehensive in function (thanks @lorcap, see https://github.com/micropython/micropython-lib/pull/449). There is a `.patch` already in the source tree to re-enable naive datetime support on platforms with meaningful `time.localtime()` :heart: .

(An aside, CPython's `time` module behaviour on Linux was a little different to my initial naive expectation. `time.timezone`, `time.alttime`, `time.dst` variables are writable but changing them doesn't change the `time.localtime()` result. To do this requires updating `TZ` (either locally or globally) and calling `time.tzset()`, which then updates those three variables to match the result of `TZ`. This makes sense (otherwise what happens when DST transitions and a CPython process is running?) but I still had naively expected changing the variables to have some effect.)

### Existing port support

* As mentioned, the Unix port already has an implementation of `time.localtime()` that calls the POSIX `localtime()` function. It could easily add more timezone support in the same way (i.e. `tzset()`), if there was a use for that.
* On some other platforms (Windows, [ESP32](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/system_time.html)) the platform has support for POSIX timezone specifications but MicroPython doesn't use them. (There may be tradeoffs on some platforms, for example ESP-IDF has full support for `TZ` environment variable and setting it but I don't know how much binary size this adds.) Some work has been done on these already, for example @tve has [this PR for ESP32](https://github.com/micropython/micropython/pull/5973).
* On most other platforms there is zero existing support for any of this.

## Way forward

This is just me spitballing at the moment:

* For esp32 port, it's technically possible to add a fully CPython compatible implementation. The [existing PR](https://github.com/micropython/micropython/pull/5973) is almost exactly this, except it adds a (helpful but incompatible) timezone argument to `time.tzset()` rather than requiring the `TZ` environment variable to be set. (Currently only Unix port has a Python interface for setting environment variables, but we could also cheat and expose something like `esp32.TZ`.)
* For bare metal ports, it seems like it'd be viable and lightweight to have something like `machine.RTC.local_tz` where a timezone offset in seconds can be written (maybe even `micropython.local_tz`?). If the application sets this, `time.localtime()` applies this offset. (The obvious choice here would be to implement `time.timezone`, etc. and use those, but that isn't how CPython uses them. :crying_cat_face:)
* There is also an idea proposed by @MrEbbinghaus ([see here](https://github.com/micropython/micropython-lib/pull/715#issuecomment-1706544340)) that ports could choose between supporting simple timezone strings with an offset (i.e. `CET-1`), or supporting the full POSIX timezone strings (`CET-1CEST,M3.5.0,M10.5.0/3`).
* Micropython 2.0 could make a breaking change and return either a 9-tuple or `struct_time` from `gmtime()` and `localtime()`, making the results compatible to CPython.
* (Getting amibitious) For the true timezone enthusiast, it would be awesome to `mip install micropython-timezone` (or similar) that adds a way to set a POSIX timezone specifier string which is then parsed in Python and automatically applied to the system. This would require adding some kind of hook to `time.localtime` or having the micropython-lib package monkey patch that function, though, and I don't know if that would be acceptable.

**Comment by projectgus** [comment]
> This would require adding some kind of hook to time.localtime or having the micropython-lib package monkey patch that function, though, and I don't know if that would be acceptable.

Maybe this isn't true actually, if the `local_tz` variable is already implemented then all this might need is a way to install a callback from machine.RTC to rewrite `local_tz` when the timezone transitions...

**Comment by ThinkTransit** [comment]
I personally use the datetime module because it supports creating tz objects with daylight savings support etc.... I like the concept of using the datetime module because it is compatible with cpython however I feel it is a little bit slow, when compared to doing a simple add/subtract of seconds.

You mention that ```None of CPython's [time.timezone, time.tzname, time.tzset(), etc.](https://docs.python.org/3.5/library/time.html#time.timezone) are implemented in MicroPython due to (reasonable) size constraints.```
Would it be a feasible to implement these but have them as optional module/component?

**Comment by Wind-stormger** [comment]
Adding the least amount of code  [#695 ntptime.py#L45-L46](https://github.com/Wind-stormger/micropython-lib/blob/2b2c7875670030612d6d33dcca9ff6560c6be8e0/micropython/net/ntptime/ntptime.py#L45-L46) is my idea, which is also derived from [peterhinch/ntptime.py#L36C19](https://github.com/peterhinch/micropython-samples/blob/56b06d26e98907af51893b9c538d667949b60f61/ntptime/ntptime.py#L36C19-L36C19).

It seems like there are some drawbacks to this? 

I mean, will this cause problems in practical application?

During my test, after setting the time zone, the `time.gmtime()` return was consistent with my local time.

**Reply by projectgus** [reply]
> During my test, after setting the time zone, the time.gmtime() return was consistent with my local time.

You mean `time.gmtime()` returned the local time and not UTC time, is that right? I agree this probably works fine for a lot of applications that only need local time, but this is also the problem.

Specifically, if you include any code that relies on getting a valid Epoch timestamp from `time.time()`, or a UTC time value from `time.gmtime()`, then that code will no longer be reading the correct time.

**Reply by Wind-stormger** [reply]
So it's best to have a `time.localtime()` .

**Comment by peterhinch** [comment]
There are two application cases. [EDITED]
 1. Where the application will always run in one location. It is easy to write region-specific code which checks the current date against the rules for DST and adjusts the time accordingly.
 2. Where an application is to run in global locations then full POSIX parsing is the best solution.

As an example of 1. see [this module](https://github.com/peterhinch/micropython-samples/blob/master/date/DATE.md) which is a minimal `Date` class for doing date calculations. The `time_offset` method is written for UK DST but could readily be adapted for other locations.

**Reply by bellingeri** [reply]
In my opinion, I think it's easier to use a standard method that just asks for information about the rules (like a POSIX string) than to write/edit a piece of code...

**Reply by beetlegigg** [reply]
I like your Date class.  All my date calcs will only be for one timezone and your date class is simple and sweet and its now going into all my code that needs to deal with time.   For long running programs I was just thinking as to how often it should get a new time from a time server to keep the microcontroller at least accurate to the minute as looking at one of my long running mqtt_as programs its seems to have slipped some 30 ish seconds over about six months.  I think getting it to grab a new time reading on the first day of each month will keep it all ticking along nicely.

**Reply by projectgus** [reply]
I agree in general @peterhinch and @bellingeri, although I note there's a special case of "1" where a significant proportion of people are in timezones with no DST - so that option is really easy for them, and less easy for the rest of us.

Suggest we can break down the use cases further as:

1. Application doesn't care about local time. (Also, may prefer that timezone handling doesn't take up any extra resources.)
2. Application has a single known timezone. The timezone has no DST, or the application doesn't care about DST. They can set an offset to UTC one time, job done! I think micropython should probably support this out of the box, or with a very small micropython-lib package.
3. Application has a single known timezone. The timezone has DST. The application wants a minimal resource footprint. They can save resources and write custom code for their particular DST transitions. If they want `localtime()` to keep working over time then this can modify the UTC offset (or set a dst flag) when it changes.
4. The application doesn't know what timezone it will run in, or does know and wants it to "just work". They install "full" timezone support (via micropython-lib) and provide a POSIX timezone string (or allow it to be configured somewhere in the application). localtime() now works the same as on a "big" computer.

> For long running programs I was just thinking as to how often it should get a new time from a time server to keep the microcontroller at least accurate to the minute 

This is an interesting question, but please start another discussion to ask it as the answer is a "whole thing" that can depend on a few factors (all unrelated to timezones). :smile:

**Reply by MrEbbinghaus** [reply]
You can do 2 with the POSIX string as well. The timezone string isn't a timezone by itself. 
The user would have to get the TZ string for the location and call the `tz_set` function regularly. 

```python
TZ_OFFSET_s = 0
FORMAT1 = re.compile(r"^([A-Z]+)([+-]?)(\d+)(?::(\d+))?(?::(\d+))?")

def offset_from_format1(tz_str) -> int:
  tzname, sign, d_hour, d_min, d_sec = re.match(FORMAT1, tz_str).groups(default=0)
  offset = (int(d_hour) * 60 + int(d_min)) * 60 + int(d_sec)
  timezone = offset * (1 if sign == "-" else -1)
  return timezone
  
def dst_data(tz_str: str, year: int) -> tuple[int, int, int]:
  pass # TODO

def offset_from_format2(tz_str):
  non_dst_offset = offset_from_format1(tz_str)
  current_time = time.time() + non_dst_offset
  current_year = time.gmtime(current_time)[0]
  dst_start, dst_end, dst_offset = get_dst_data(tz_str, current_year)
  
  return current_time + (dst_offset if dst_start < current_time < dst_end 0)
  

def tz_set(tz_str = None):
  global TZ_OFFSET_s
  if tz_str:
    TZ_OFFSET_s = offset_from_format2(tz_str)
```

--- 

"full _just works_ TZ support" is a little bit harder, I'm afraid.

You would need to regularly (at least twice a year) update your timezone database, as DST rules could change any moment.

**Reply by projectgus** [reply]
> "full just works TZ support" is a little bit harder, I'm afraid.
> You would need to regularly (at least twice a year) update your timezone database, as DST rules could change any moment.

Good point, I shouldn't have oversimplified. At best, "just works" really only extends as far as "an application can take the POSIX timezone string that applies for their current time zone and pass it to some API in order to get the corresponding local time behaviour."

I don't think there's a possible outcome where MicroPython starts including an actual timezone database.

**Reply by bellingeri** [reply]
Did you find problems with my solution ([localPTZtime](https://github.com/bellingeri/localPTZtime))?

**Reply by projectgus** [reply]
@bellingeri I don't know if you're asking me or @MrEbbinghaus but speaking for myself I don't see any technical problem with it at all (although i haven't looked very closely).

As to the question of implementing an official MicroPython solution then there are two other factors to using it:

1. Where possible MicroPython aims to provide the same APIs as CPython.
2. localPTZtime is licensed GPLv3, which is a perfectly valid choice but this can't be included (or used as a basis for new work) in MicroPython (which is MIT licensed).

**Comment by cwalther** [comment]
One detail I haven’t seen mentioned yet: Since currently `time.gmtime` and `time.localtime` are the same thing, I can misuse `time.mktime` as a substitute for `calendar.timegm`. If `time.localtime` and its inverse `time.mktime` were changed to actually use local time, we would also need an implementation of `calendar.timegm` to have an inverse of `time.gmtime`.

**Reply by projectgus** [reply]
Huh... I actually didn't know `calendar.timegm` was a thing, but it makes sense that this function exists! I guess the other way around this is to construct a `datetime` object in UTC and read its timestamp, but that's fiddlier. Definitely something to keep in mind.

One good thing about all the proposed changes, as I understand them, is that anyone should have the option to keep `localtime` in UTC and continue with whatever techniques work for them now.

**Comment by scruss** [comment]
>     1. Where the application will always run in one country. It is easy to write country-specific code which checks the current date against the rules for DST and adjusts the time accordingly.

Please don't generalize for simple countries. Consider Canada:

* we have at least six time zones, from Pacific (currently GMT-7) to Atlantic (currently GMT-3) — and also Newfoundland (currently GMT-2:30);
* not every province/territory observes DST: *most* of Saskatchewan doesn't, so during the summer it has the same time as Alberta to the west of it, but shares the winter time with Manitoba to the east;
* some provinces have more than one time zone. Ontario, for example, is mostly in Eastern time, but the far NW is Central. Other provinces have splits not based strictly on longitude but on local cultural preference.

It's very complicated and very confusing: [Time zones and daylight saving time - National Research Council Canada](https://nrc.canada.ca/en/certifications-evaluations-standards/canadas-official-time/time-zones-daylight-saving-time).

**Reply by peterhinch** [reply]
Good point. I have corrected my post.

**Comment by kjm1102** [comment]
If the shakers & movers are going to make local time easily available in micropython it might be worth fixing some of the other time related annoyances? In particular the inconsistency between datetime (which uses the year, month, day-of-month, day-of-week, hours, mins, secs, day-of year format) and  mktime or gmtime (which use the  year, month, day-of-month, hour, mins, secs, day-of-week, day-of-year format).

Interestingly, to me at least, if your project uses cellular connectivity most LTE modems will supply local time via the 4G tower. Even then you are not out of the woods when your application blunders into the twilight zone around 3am twice a year when a whole hour of local time either vanishes or pops into existence like some magical sub atomic particle.

**Reply by projectgus** [reply]
> In particular the inconsistency between datetime (which uses the year, month, day-of-month, day-of-week, hours, mins, secs, day-of year format) and mktime or gmtime (which use the year, month, day-of-month, hour, mins, secs, day-of-week, day-of-year format).

Interesting, do you have a specific example or a link to where this is discussed in more detail? AFAIK all of the micropython-lib datetime APIs match CPython (aside from the gmtime() different mentioned in the first post). However those APIs don't quite match this description.

 If there's somewhere that the behaviour either doesn't match CPython, or is a MicroPython-only thing, then there might be room to change it.

> Interestingly, to me at least, if your project uses cellular connectivity most LTE modems will supply local time via the 4G tower. Even then you are not out of the woods when your application blunders into the twilight zone around 3am twice a year when a whole hour of local time either vanishes or pops into existence like some magical sub atomic particle.

That is indeed interesting! Do the LTE modems give you the option of UTC, at least?

It must also get weird if you live in an area close to a timezone border, and are able to reach towers in more than one timezone...

**Reply by ThinkTransit** [reply]
One example here is that datetime.now() isn't fully implemented and returns "not implemented" unless passed a tz object. You could argue that now() is one of the most common datetime functions and not having a working now() doesn't instil confidence in the datetime module. Hoping that this can be merged soon to rectify this.

https://github.com/micropython/micropython-lib/pull/570

**Comment by bixb922** [comment]
> For esp32 port, it's technically possible to add a fully CPython compatible implementation.

This is indeed interesting and would be a good optimization, making the ESP32 implementation smaller. However, it's probably more compatible and less maintenance effort to have the same implementation for all architectures.

>For bare metal ports, it seems like it'd be viable and lightweight to have something like machine.RTC.local_tz where a timezone offset in seconds can be written

Having three variables (time zone name, offset and DST indicator) and setting these variables sounds like a simple solution. To use that, the programmer will have to write a task that wakes up when a transition between DST and non-DST occurs, and set the new variables. However, there is no guarantee that this task, be it asyncio, timer or thread, will wake up exactly at the specified time. This task to set the time zone may be triggered a bit late, and other code can call ```time.localtime()``` in between. So there will be an interval of time where the offset should have changed, but hasn't yet. This is not desirable. One way to avoid this little design problem is to provide a callback that is called each time time.localtime() is invoked, and that returns (zone name, time offset, DST indicator). I believe the overhead of this call is minimal.

In summary: it might be better to process the time zone rule each time ```time.localtime()``` gets called instead of setting time zone variables.

The same reasoning applies to an RTC. I believe the RTC should be set to UTC (never to the local time) and the time zone rules should be applied by software each time the RTC is read. There are several advantages: it makes it possible to work with several time zones, and there is no need for a time critical task to update the RTC on time zone transitions. In fact, if the RTC is set to a a certain local time, it's very tricky to move it back to UTC or to another time zone.

>or supporting the full POSIX timezone strings (CET-1CEST,M3.5.0,M10.5.0/3).

This indeed would be very useful. Due to space restraints, this functionality most likely might be mip installable as an optional module. To be compatible with big Python, ```os.environ[]``` might be used to set the TZ environment variable and ```time.tzset()``` should be required to have support for one time zone at a time.

I think having a MicroPython specific time zone data base may be a lot of work. The database has to be updated weekly or even more frequently. It's probably better to have everyone lookup or request the needed POSIX string. Updating time zone rules via mip install Continent/City) may be interesting, but need not be part of MicroPython for now. 

Just to illustate: In some countries, time zone rules have been changed on short notice (a week before the transition). This must be handled as OTA emergency patch. There are many ways available today to support OTA updates, no need to do something time zone specific.

**Reply by projectgus** [reply]
> One way to avoid this little design problem is to provide a callback that is called each time time.localtime() is invoked, and that returns (zone name, time offset, DST indicator). I believe the overhead of this call is minimal.
>
> In summary: it might be better to process the time zone rule each time time.localtime() gets called instead of setting time zone variables.

Yes, this is more or less how I imagined it would work as well. If the UTC timestamp of the next timezone offset change is stored somewhere then comparing the current time against it is a lightweight check for whether or not the offset needs to be re-calculated. ~~(localtime always resolves in the current timezone of "now", regardless of what timestamp is passed in the argument.)~~ (EDIT: not the case at least in CPython, see two replies down.)

> I think having a MicroPython specific time zone data base may be a lot of work. 

Fully agree, that is a bridge too far.

**Reply by cwalther** [reply]
> To use that, the programmer will have to write a task that wakes up when a transition between DST and non-DST occurs, and set the new variables.

Apart from the other points you mention, this is not correct to begin with. What DST offset to use on a particular`localtime(t)` call does not depend on whether DST is in effect _now_, but on whether it was/is/will be in effect _at time `t`_.

**Reply by bixb922** [reply]
@cwalther, you are completely right. What I tried to explain in my previous post is that writing a task to change the time offset by setting a variable is not desirable. I'm sorry if could not explain this well. You added yet another good argument against that design.


I had understood (perhaps wrongly) that the proposal of the OP was about timestamps, in the sense of converting "now" to local time. I think ```time.localtime()```  (i.e. getting "now" in local time) is by far the most common and needed use case.

You are also right that ```time.localtime(t)```needs more work. Since the time zone rules change in time, this requires to have a history of past and future time zone rules. In some countries time zone rules have changed every 2 to 3 years. So this could mean having, for example, 5 POSIX time zone strings for the last 10 years, each rule with the time range it was valid, plus a rule for a time zone rule change in the near future. This means additional complexity that has to be weighed against usefulness.

**Reply by projectgus** [reply]
> What DST offset to use on a particular localtime(t) call does not depend on whether DST is in effect now, but on whether it was/is/will be in effect at time t.

I'd convinced myself this was otherwise (I found the [POSIX description](https://pubs.opengroup.org/onlinepubs/9699919799/functions/localtime.html) ambiguous), but of course you're right @cwalther. In CPython, the `tm_gmtoff`, `tm_isdst`, `tm_zone` fields reflect the local timezone at time `t`.

> I had understood (perhaps wrongly) that the proposal of the OP was about timestamps, in the sense of converting "now" to local time. I think time.localtime() (i.e. getting "now" in local time) is by far the most common and needed use case.

You're totally correct that this was the motivation for looking into this, as it's by far the most common use case. However for any change in MicroPython, especially if it's an API that also exists in CPython, then we have to think through the other implications as well.

**Reply by scruss** [reply]
Talking of other implications: in today's Melbourne MicroPython Meetup, Damien noted that with the recent addition of [SSL/TLS support to asyncio](https://github.com/micropython/micropython/releases/tag/v1.22.0), it's important to have your MicroPython board set to the correct time. An SSL connection will fail if the system's clock is wrong. It was suggested to use `mpremote rtc --set` to correct this, which sets the RTC to *local* time.

**Reply by clach04** [reply]
@scruss I think you were clear but I want to be as clear as possible, rather than infer 😄 . Are you saying that to make use of TLS/SSL, devices running MicroPython need to be set to local time and not UTC time? I've not had chance to play with the SSL lib on MicroPython yet.

**Reply by clach04** [reply]
I tested requests (not async) and it doesn't seem to care how wrong the clock is with https connections 😢 I've not had chance to dig any deeper into this.

**Comment by viji2ks** [comment]
guys here first most impordent is to connect board to wifi n internet using network commands then this is simple micropython code to see and sync ur current boad to network time Njoy
plz save below as getNow.py and rest how to excute see the code below last line 


import time
import ntptime
try:
    ntptime.settime()
except:
    raise Exception("npttime.settime() failed. No network connection.")

def getNow(UTC_OFFSET=+5.5):
    local_time = int(UTC_OFFSET * 3600)
    return time.gmtime(time.time() + local_time)

if __name__ == "__main__":
    print(getNow())
    
    
"""
getNow.py
viji2ks@gmail.com
2022-12-12

Usage (assuming a it is connected to wifi with internet network connection):
>>> from getNow import getNow
>>> print(getNow(UTC_OFFSET=-5))
(2022, 12, 12, 10, 34, 34, 0, 346)

"""

**Comment by createdbyjurand** [comment]
If the problem is that there is no marker for the time zone, you could parse `datetime()` to ISO UTC marking it with "zulu" GMT timezone and all other systems should be able to convert it to local time.

`time.time()` is unfortunately less precise because it lacks milliseconds count. 

```python
ntptime.settime()

rtc = machine.RTC()

def dateTimeToISO(dateTime):
    return "{:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}.{:03d}Z".format(
        dateTime[0],
        dateTime[1],
        dateTime[2],
        dateTime[4],
        dateTime[5],
        dateTime[6],
        dateTime[7] // 1000,
    )

dateTimeToISO(rtc.datetime())
```

**Comment by sosi-deadeye** [comment]
To use timezones, the `ZoneInfo` class should be implemented for Micropython. **But you cannot put all tzdata on a Microcontroller, because it's huge.** 

```
(TZ) [deadeye@nexus zoneinfo]$ du -sh /usr/share/zoneinfo
4,5M    /usr/share/zoneinfo
(TZ) [deadeye@nexus zoneinfo]$ du -sh /usr/share/zoneinfo/Europe/
160K    /usr/share/zoneinfo/Europe/
```
And the package tzdata could also be used to get the timezones, but the size is different:
```
(TZ) [deadeye@nexus zoneinfo]$ du -sh ../zoneinfo/
2,8M    ../zoneinfo/
(TZ) [deadeye@nexus zoneinfo]$ du -sh Europe/
268K    Europe/
```

For every political change to the time zone, one entry extends the file. There are no rules, but rather points in time that then define the offset. In extreme cases, there were once 150 time zone changes in just one year.

But due to the directory structure, you can leave out the time zones that are not used in advance. For example, if I only use something in Europe, I don't need the time zones of the USA.

EDIT: On Linux-Systems, the `TZif2` format is used to store the zoneinfo. The format is described here: https://www.rfc-editor.org/rfc/rfc8536.html#page-8

**Comment by clach04** [comment]
My personal/selfish use case is a simple clock which needs to have "now" information and (near) future. I do not have a need for historically accurate (i.e. previous) local values. I do want more than one timezone to be usable. That makes my use case a lot more simple (albeit not trival) 😃 

I really like the idea that @MrEbbinghaus and @bellingeri discussed, which is to use POSIX TZ names and rules (NOT IANA). Again this happens to fit my personal use case, it may not match anyone else's! 😁 I'm OK updating config with a new rule should politics change in the area(s) of interest.

  * @MrEbbinghaus had a partial skeleton of code at https://github.com/orgs/micropython/discussions/12378#discussioncomment-6949368
  * @bellingeri has a library https://github.com/bellingeri/localPTZtime which didn't really work out for me, I'm not keen on relying on float for time when MicroPython uses integer. Also it seems to parse the TZ on each execution call.
  * @divergentti had a similar but different approach, his use case appears to be similar to mine in terms of needs. He posted to https://github.com/orgs/micropython/discussions/11173#discussioncomment-6059669 and has code at https://github.com/divergentti/Micropython-for-ESP32/blob/346f35d99699423be7d2bd10891b2a8855b262a5/Positioning/esp32-gps-oled-bme680/main.py#L59

So I had a case of [NIH](https://en.wikipedia.org/wiki/Not_invented_here) and created https://github.com/clach04/py-posix_tz - it is super restrictive, it doesn't work the same under CPython  (and I'm OK with that 😆) It's not been benchmarked, and I don't know how much memory is uses yet (I need something tiny as I keep running out of memory).

There is a minuscule test suite, it is **very incomplete** and rough around the edges, and has a very poor manual cache. But it looks like it will work for me.

Kudos to @bellingeri and @MrEbbinghaus for proposing TZ, with IANA being my day-to-day I'd almost forgotten about posiz TZ rules, I can use this approach (but not code) for a different embedded project (wrist watch) where I'm trying to keep storage to a minimum.

There are a bunch of tools out there to generate TZ rules from IANA if you don't want to manually create a rule.

**Reply by bulletmark** [reply]
As I have written elsewhere here, I have an application template/framework which uses the standard MicroPython [datetime](https://github.com/micropython/micropython-lib/blob/master/python-stdlib/datetime/datetime.py) + [`class timezone(tzinfo)`](https://github.com/micropython/micropython-lib/blob/68e3e07bc7ab63931cead3854b2a114e9a084248/python-stdlib/datetime/datetime.py#L247C1-L247C22) implementation so I instantiate that using a user configured timezone offset (per device). The user can optionally set that to  a string, i.e. a zone name, e.g. "Australia/Sydney". My ESP32 application framework uses an async background service to periodically sync the clock to UTC time using [`ntptime.settime()`](https://docs.micropython.org/en/latest/esp8266/quickref.html#real-time-clock-rtc) and I get the offset for timezone name if specified from a free API such as https://www.timeapi.io/ or http://worldtimeapi.org/ (at a slower rate). This is a simple approach but also handles political, DST, and any other changes.
